File: app/src/main/java/coolpharaoh/tee/speicher/tea/timer/core/counter/Counter.java
package coolpharaoh.tee.speicher.tea.timer.core.counter;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.Ignore;
import androidx.room.Index;
import androidx.room.PrimaryKey;
import androidx.room.TypeConverters;

import java.util.Date;
import java.util.Objects;
import java.util.stream.Stream;

import coolpharaoh.tee.speicher.tea.timer.core.date.DateConverter;
import coolpharaoh.tee.speicher.tea.timer.core.tea.Tea;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@NoArgsConstructor
@Entity(tableName = "counter", foreignKeys = @ForeignKey(entity = Tea.class, parentColumns = "tea_id", childColumns = "tea_id", onDelete = ForeignKey.CASCADE), indices = {@Index("tea_id")})
public class Counter {
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "counter_id")
    private Long id;

    @ColumnInfo(name = "tea_id")
    private long teaId;

    @ColumnInfo(name = "week")
    private int week;

    @ColumnInfo(name = "month")
    private int month;

    @ColumnInfo(name = "year")
    private int year;

    @ColumnInfo(name = "overall")
    private long overall;

    @TypeConverters(DateConverter.class)
    @ColumnInfo(name = "week_date")
    private Date weekDate;

    @TypeConverters(DateConverter.class)
    @ColumnInfo(name = "month_date")
    private Date monthDate;

    @TypeConverters(DateConverter.class)
    @ColumnInfo(name = "year_date")
    private Date yearDate;

    // the constructor needs these parameters
    @SuppressWarnings("java:S107")
    @Ignore
    public Counter(final long teaId, final int week, final int month, final  int year, final long overall,
                   final Date weekDate, final Date monthDate, final Date yearDate) {
        this.teaId = teaId;
        this.week = week;
        this.month = month;
        this.year = year;
        this.overall = overall;
        this.weekDate = weekDate;
        this.monthDate = monthDate;
        this.yearDate = yearDate;
    }

    public boolean hasEmptyFields() {
        return Stream.of(weekDate, monthDate, yearDate).anyMatch(Objects::isNull);
    }
}

File: app/src/main/java/coolpharaoh/tee/speicher/tea/timer/core/counter/Counter.kt
package coolpharaoh.tee.speicher.tea.timer.core.counter

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.ForeignKey.Companion.CASCADE
import androidx.room.Ignore
import androidx.room.Index
import androidx.room.PrimaryKey
import androidx.room.TypeConverters
import coolpharaoh.tee.speicher.tea.timer.core.date.DateConverter
import coolpharaoh.tee.speicher.tea.timer.core.tea.Tea
import java.util.Date
import java.util.Objects
import java.util.stream.Stream

@Entity(
    tableName = "counter",
    foreignKeys = [ForeignKey(
        entity = Tea::class,
        parentColumns = arrayOf("tea_id"),
        childColumns = arrayOf("tea_id"),
        onDelete = CASCADE
    )],
    indices = [Index("tea_id")]
)
data class Counter(
    @ColumnInfo(name = "counter_id") @PrimaryKey(autoGenerate = true) var id: Long? = null,
    @ColumnInfo(name = "tea_id") var teaId: Long = 0,
    @ColumnInfo(name = "week") var week: Int = 0,
    @ColumnInfo(name = "month") var month: Int = 0,
    @ColumnInfo(name = "year") var year: Int = 0,
    @ColumnInfo(name = "overall") var overall: Long = 0,
    @ColumnInfo(name = "week_date") @field:TypeConverters(DateConverter::class)
    var weekDate: Date? = null,
    @ColumnInfo(name = "month_date") @field:TypeConverters(DateConverter::class)
    var monthDate: Date? = null,
    @ColumnInfo(name = "year_date") @field:TypeConverters(DateConverter::class)
    var yearDate: Date? = null
) {

    @Ignore
    constructor(teaId: Long, week: Int, month: Int, year: Int, overall: Long, weekDate: Date?,
                monthDate: Date?, yearDate: Date?) :
            this(null, teaId, week, month, year, overall, weekDate, monthDate, yearDate)

    fun hasEmptyFields(): Boolean {
        return Stream.of(weekDate, monthDate, yearDate)
            .anyMatch { obj: Date? -> Objects.isNull(obj) }
    }
}